prov/util/src/util_mr_cache.c

```diff
@@ -295,6 +300,18 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
 		goto unlock;
 	}
 
+	if (((*entry)->access & attr->access) != attr->access) {
+		util_mr_uncache_entry(cache, *entry);
+		access = attr->access | (*entry)->access;
+		if ((*entry)->use_cnt == 0) {
+			dlist_remove_init(&((*entry)->lru_entry));
+			util_mr_free_entry(cache, *entry);
+		}
+		ret = util_mr_cache_create(cache, attr->mr_iov, access, entry);
+		goto unlock;
+
+	}
```

Comment:
This is bypassing the merge region handling (util_mr_cache_merge), which is an issue.  The entry returned by the find() callback may be larger than the new region or shifted left/right.  The new entry that's created here combines the access permissions, but the size is the same as the new registration.

If we're supposed to be merging regions, then there should not be overlapping regions in the cache.  But because this skips the merge region processing, it could end up in that situation.  Also, if the access does match, then we can fall through to the util_mr_cache_merge() call, which will blindly merge the new regions, without adjusting for the access permissions.

We need to handle access differences as part of util_mr_cache_merge().  So the functionality above needs to move inside that call somehow.  Additionally, we need to handle this as part of an else clause to the if statement below (reproduced here):
```
if (!ofi_iov_within(attr->mr_iov, &(*entry)->iov)) {
    ret = util_mr_cache_merge(cache, attr, *entry, entry);
    goto unlock;
} else if (((*entry)->access & attr->access) != attr->access) {
    /* new code from above */
    ...
    /* Use existing entry for iov, new region is within it */
    ret = util_mr_cache_create(cache, &(*entry)->attr.iov, access, entry);
    goto unlock;
}
```
I would start by updating the merge() call.  That may produce common code that can be pulled out and re-used in the else if clause.

Reply:
Please see latest updates. Added a patch that updates the uncache function to remove entry from LRU list as well if use_cnt is 0. Added the check for access mismatch simply just after the "merge needed check". This should be fine as the merge function handles access mismatch as well.